### PR TITLE
[stakingindex] implement indexer for new staking contract

### PR DESCRIPTION
**Pull request by envestcc**
# Description
The new version of the staking contract is coming, mainly removing the restrictions on the staking amount, embracing liquidity-hub better. Therefore a new indexer is also required. 

Three main changes:
- add  a common pkg `pkg/util/abiutil` to handle data structure convertion between contract and go
- add indexer common to implement some basic for all contract indexers
- add the new staking indexer for the new staking contract

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: systemcontractindex/stakingindex/index.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	bkt := s.cache.Bucket(id)
+	if bkt == nil {
+		return nil, false, nil
+	}
+	vbs := assembleVoteBucket(id, bkt, s.common.ContractAddress(), s.blockInterval)
+	return vbs, true, nil
+}
+
+// BucketsByIndices returns the buckets by indices
+func (s *Indexer) BucketsByIndices(indices []uint64, height uint64) ([]*VoteBucket, error) {
```
Comment (CoderZhi) [inline systemcontractindex/stakingindex/index.go:106]: `Buckets`

File: systemcontractindex/stakingindex/bucket.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	bi.Candidate = candidate
+	bi.Owner = owner
+	bi.StakedAmount = amount
+	bi.StakedDurationBlockNumber = p.Duration
+	return nil
+}
+
+func (b *Bucket) Clone() *Bucket {
+	clone := &Bucket{
+		StakedAmount:              b.StakedAmount,
```
[the remark this answers is not in the document]
Comment (envestcc) [reply systemcontractindex/stakingindex/bucket.go:80]: it has been used in fetching methods to return copied buckets

File: systemcontractindex/common.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// StartHeight returns the start height of the indexer
+func (s *IndexerCommon) StartHeight() uint64 { return s.startHeight }
+
+// PutHeight puts the tip block height
+func (s *IndexerCommon) PutHeight(height uint64) error {
+	return s.kvstore.Put(s.ns, s.key, byteutil.Uint64ToBytesBigEndian(height))
+}
+
+// BlockContinuity checks the block continuity
+func (s *IndexerCommon) BlockContinuity(height uint64) (existed bool, err error) {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply systemcontractindex/common.go:67]: okay

File: systemcontractindex/stakingindex/cache.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func (s *cache) Bucket(id uint64) *Bucket {
+	if bkt, ok := s.buckets[id]; ok {
+		return bkt
+	}
+	return nil
+}
+
+func (s *cache) BucketsByIndices(indices []uint64) []*Bucket {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply systemcontractindex/stakingindex/cache.go:118]: fixed

File: systemcontractindex/stakingindex/event_handler.go [hunk 5]
```diff
@@ -308,3 +312,13 @@ func (eh *eventHandler) Finalize() (batch.KVStoreBatch, *cache) {
...
 }
+
+func (eh *eventHandler) putBucket(id uint64, bkt *Bucket) {
+	eh.dirty.PutBucket(id, bkt)
+	eh.delta.Put(stakingBucketNS, byteutil.Uint64ToBytesBigEndian(id), bkt.Serialize(), "failed to put bucket")
+}
+
+func (eh *eventHandler) delBucket(id uint64) {
+	eh.dirty.DeleteBucket(id)
+	eh.delta.Delete(stakingBucketNS, byteutil.Uint64ToBytesBigEndian(id), "failed to put bucket")
```
Comment (CoderZhi) [inline systemcontractindex/stakingindex/event_handler.go:323]: `failed to delete bucket`

**Review by CoderZhi — APPROVED** [review]

File: systemcontractindex/stakingindex/bucket.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply systemcontractindex/stakingindex/bucket.go:79]: didn't get this, L90 is the correct way?

File: systemcontractindex/stakingindex/bucket.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		CreatedAt:                 b.CreatedAt,
+		UnlockedAt:                b.UnlockedAt,
+		UnstakedAt:                b.UnstakedAt,
+	}
+	candidate, _ := address.FromBytes(b.Candidate.Bytes())
+	clone.Candidate = candidate
+	owner, _ := address.FromBytes(b.Owner.Bytes())
+	clone.Owner = owner
+	stakingAmount := new(big.Int).Set(b.StakedAmount)
+	clone.StakedAmount = stakingAmount
```
Comment (dustinxie) [inline systemcontractindex/stakingindex/bucket.go:90]: `clone.StakedAmount = new(big.Int).Set(b.StakedAmount)`?

File: systemcontractindex/common.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	return height, nil
+}
+
+// StartHeight returns the start height of the indexer
+func (s *IndexerCommon) StartHeight() uint64 { return s.startHeight }
+
+// Commit commits the height to the indexer
+func (s *IndexerCommon) Commit(height uint64, delta batch.KVStoreBatch) error {
+	s.height = height
```
Comment (dustinxie) [inline systemcontractindex/common.go:87]: this should be the last statement after `WriteBatch()` returns nil? If write batch failed, should not update height

File: systemcontractindex/stakingindex/bucket.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
+package stakingindex
```
Comment (dustinxie) [inline systemcontractindex/stakingindex/bucket.go:1]: these new codes are same as the existing staking contract indexer?
If that's the case, we should delete the existing code?

File: systemcontractindex/common.go (hunk 7 above)
Comment (envestcc) [reply systemcontractindex/common.go:87]: u are right

File: systemcontractindex/stakingindex/bucket.go (hunk 8 above)
Comment (envestcc) [reply systemcontractindex/stakingindex/bucket.go:1]: existed staking indexer will be refacted by the new code in future

File: systemcontractindex/stakingindex/bucket.go [hunk 9]
```diff
@@ -86,8 +85,7 @@ func (b *Bucket) Clone() *Bucket {
 	clone.Candidate = candidate
 	owner, _ := address.FromBytes(b.Owner.Bytes())
 	clone.Owner = owner
-	stakingAmount := new(big.Int).Set(b.StakedAmount)
-	clone.StakedAmount = stakingAmount
+	clone.StakedAmount = new(big.Int).Set(b.StakedAmount)
```
Comment (dustinxie) [inline systemcontractindex/stakingindex/bucket.go:88]: move to L79:
`StakedAmount :  new(big.Int).Set(b.StakedAmount),`

**Review by dustinxie — APPROVED** [review]

**Pull request merged**